cpp/src/io/json/json_tree.cu

```diff
@@ -254,12 +320,485 @@ tree_meta_t get_tree_representation(device_span<PdaTokenT const> tokens,
+{
+  CUDF_FUNC_RANGE();
+  // 1. Convert node_category+fieldname to node_type.
+  nvtxRangePushA("node_type");
+  auto num_nodes = d_tree.node_categories.size();
+  rmm::device_uvector<size_type> node_type =
+    hash_node_type_with_field_name(d_input, d_tree, stream);
+  nvtxRangePop();
+
+  // TODO two-level hashing:  one for field names
+  // and another for {node-level, node_category} + field hash for the entire path
+
+#ifdef NJP_DEBUG_PRINT
+  print_vec(node_type, stream, "node_type");
+#endif
+
+  // 2. Preprocessing: Translate parent node ids after sorting by level.
+  //   a. sort by level
+  //   b. get gather map of sorted indices
+  //   c. translate parent_node_ids to sorted indices
+
+  nvtxRangePushA("parent_indices");
+  rmm::device_uvector<size_type> scatter_indices(num_nodes, stream);
+  thrust::sequence(rmm::exec_policy(stream), scatter_indices.begin(), scatter_indices.end());
+#ifdef NJP_DEBUG_PRINT
+  printf("\n");
+  print_vec(scatter_indices, stream, "gpu.node_id");
+  print_vec(d_tree.parent_node_ids, stream, "gpu.parent_node_ids");
+  print_vec(node_type, stream, "gpu.node_type");
+  print_vec(d_tree.node_levels, stream, "gpu.node_levels");
+#endif
+  rmm::device_uvector<NodeIndexT> parent_node_ids(d_tree.parent_node_ids, stream);  // make a copy
+  auto out_pid =
+    thrust::make_zip_iterator(scatter_indices.data(), parent_node_ids.data(), node_type.data());
+  // Uses cub radix sort.
+  thrust::stable_sort_by_key(rmm::exec_policy(stream),
+                             d_tree.node_levels.data(),
+                             d_tree.node_levels.data() + num_nodes,
+                             out_pid);
+
+  rmm::device_uvector<NodeIndexT> parent_indices =
+    translate_sorted_parent_node_indices(scatter_indices, parent_node_ids, stream);
+  nvtxRangePop();
+#ifdef NJP_DEBUG_PRINT
+  printf("\n");
+  print_vec(scatter_indices, stream, "gpu.node_id");
+  print_vec(parent_node_ids, stream, "gpu.parent_node_ids");
+  print_vec(node_type, stream, "gpu.node_type");
+  print_vec(d_tree.node_levels, stream, "gpu.node_levels");
+  print_vec(parent_indices, stream, "parent_indices");
+#endif
+  // 3. Find level boundaries.
+  nvtxRangePushA("level_boundaries");
+  std::vector<size_type> level_boundaries = [&]() {
+    // Already node_levels is sorted
+    auto max_level = d_tree.node_levels.back_element(stream);
+    rmm::device_uvector<size_type> level_boundaries(max_level + 1, stream);
+    // TODO try reduce_by_key
+    auto level_end =
+      thrust::copy_if(rmm::exec_policy(stream),
+                      thrust::make_counting_iterator<size_type>(1),
+                      thrust::make_counting_iterator<size_type>(num_nodes + 1),
+                      level_boundaries.begin(),
+                      [num_nodes, node_levels = d_tree.node_levels.begin()] __device__(auto index) {
+                        return index == num_nodes || node_levels[index] != node_levels[index - 1];
+                      });
```

Comment:
maybe this would be clearer as a  `thrust::unique_copy`?

Reply:
`thrust::unique_copy` will copy first unique element. I want to copy the last unique element here.